### ISSUE TITLE
Migration from fp-ts: replacements for fp-ts-routing and hyper-ts

### ISSUE THREAD

**Issue by OliverJAsh** [Effect-TS/effect#2203]
(cc @schickling who told me to open this issue. 👋)

At [Unsplash](https://unsplash.com/) we are heavily using fp-ts and the surrounding ecosystem, namely the [fp-ts-routing](https://github.com/gcanti/fp-ts-routing) and [hyper-ts](https://github.com/DenisFrezzato/hyper-ts) extension libraries.

We are keen to migrate over to Effect however this would require replacements for fp-ts-routing and hyper-ts.

fp-ts-routing is lacking documentation but hopefully my blog post can shed light on why we find it to be so valuable at Unsplash: https://oliverjash.me/2022/type-safe-routing-in-react-with-fp-ts-routing-part-1.

Are there any existing libraries that would fill these gaps? If not, could we collaborate to make this happen? ❤️

**Comment by jessekelly881** [comment]
You should be able to replace hyper-ts with @effect/platform/HttpServer. I'm not aware of any routing solutions at the moment though. I'd love to see something like tanstack router on top of @effect-rx though.

**Comment by OliverJAsh** [comment]
What I like about fp-ts-routing:

- It’s very simple: it’s just a string parser and formatter. I think of it as similar to io-ts/schema in that it’s a bidirectional transformation, from a string to a sum type and back.
- It encourages you to model your routes as a sum type which is easy to match against.
- You get link builders for free.
- Representing parsers as a monad makes composition very easy. This is one of the reasons we adopted fp-ts-routing at Unsplash (coming from path-to-regexp): we needed to internationalise the site and with fp-ts-routing it was extremely easy to add a locale prefix to the URL path and also to translate each part of the URL.

I don’t think I could ever go back to something like React Router or Tanstack Router.

cc @gcanti who created this amazing library. Curious to hear your thoughts about how this idea could be ported to Effect.

**Comment by TylorS** [comment]
@OliverJAsh I took the time to read the article you wrote and linked above, and it may not be exactly what you're looking for, it's definitely not a 1:1 with fp-ts-routing, and it _is_ based on `path-to-regexp`, but I have a general routing solution I've been working on with Effect. The full functionality is split across a few different modules.

Firstly, I have `@typed/path` which defines a subset of the path-to-regexp DSL, it includes everything except arbitrary regular expressions, but definitely supports optionals, prefixes, zeroOrMore, oneOrMore, and even query parameters. It has bi-directional support, so it correctly parses the path syntax to records of corresponding parameters, and it also supports type-level interpolation in the opposite direction. There's a good chance you wouldn't need to interact directly with this library, but it does serve as the base and has 0 runtime dependencies.

More directly useful is `@typed/route`, which does require `path-to-regexp` as a runtime dependency, and it provides the runtime behavior of matching paths, creating paths, concatenating them using a type called `Route<Path>`. It's intended to be useful standalone, so that's pretty much all it does. Though it does implement an `AsGuard<I, O, E, R>` interface from my `@typed/guard` package which allows extends things with Effect.

The `@typed/guard` package works around a type `Guard<I, O, E, R>` which is just an alias to `(i: I) => Effect<Option<O>, E, R>`. All `Route`s implicitly are supported in the Guard operators, which all around allow you to compose, map, filter, encode, decode, or anything an Effect can do really.

I also have `@typed/navigation`, and `@typed/router` packages to further improve how all of this comes together in a browser or server environment, but they would assume a dependency on my push-based stream abstractions from `@typed/fx` which may or may not be ideal for you.

TLDR; That's a lot of words, but it's probably easier just to see some code - https://github.com/TylorS/typed/blob/development/packages/route/test/index.ts#L76

**Comment by gcanti** [comment]
RE: `fp-ts-routing`, the codebase is small, making a 1:1 port should not be difficult

p.s.
However, I don't know if it's of interest to the effect org, /cc @mikearnaldi

**Comment by mikearnaldi** [comment]
> RE: `fp-ts-routing`, the codebase is small, making a 1:1 port should not be difficult
> 
> p.s. However, I don't know if it's of interest to the effect org, /cc @mikearnaldi

It would be ideal if the folks who need such library would sponsor the development, it is a very FP specific lib that I am not sure solves a widespread problem. Happy to help with advise/review if they want to take the lead

**Comment by OliverJAsh** [comment]
I think both of the libraries I mentioned are solving widespread problems, certainly they are problems we encounter very frequently at Unsplash which is a typical CRUD application.

Maybe we can start with fp-ts-routing?

**Comment by jessekelly881** [comment]
As a side note, a port of zio-parser would probably be a better long-term/general solution. You could easily build a bi-directional param/query parser like fp-ts-routing using it.

**Comment by mikearnaldi** [comment]
@IMax153 was playing with a zio-parser port

**Comment by IMax153** [comment]
I've sort of "abandoned" this project for now, but in case anyone is curious :) https://github.com/IMax153/effect-ts-parser

**Comment by jessekelly881** [comment]
Nice! I have my own "abandoned" port of effect-parse as well jajaja. I think your version is closer to complete though. https://github.com/jessekelly881/effect-parser
